### ISSUE TITLE
Filter by tag

### ISSUE THREAD

**Issue by vichusharma** [duo-labs/cloudmapper#36]
Provide option to Prepare data only for the Tags as it might get too complicated for an enterprise to run it on there infrastructure 

- 0 nodes built in region eu-west-3
- 0 nodes built in region eu-west-2
- 6 nodes built in region eu-west-1
- 0 nodes built in region ap-northeast-2
- 0 nodes built in region ap-northeast-1
- 0 nodes built in region sa-east-1
- 0 nodes built in region ca-central-1
- 0 nodes built in region ap-southeast-1
- 0 nodes built in region ap-southeast-2
- 0 nodes built in region eu-central-1
- 999 nodes built in region us-east-1
- 0 nodes built in region us-east-2
- 0 nodes built in region us-west-1
- 0 nodes built in region us-west-2
- 507426 connections built
WARNING: There are 1120 total nodes and 507426 total edges.
  This will be difficult to display and may be too complex to make sense of.
  Consider reducing the number of items in the diagram by viewing a single
   region, ignoring internal edges, or other filtering.

**Comment by rafarost** [comment]
That would be helpful. I was facing the same issue here. I started looking into the code, and realized that AWS CLI does not return the tags. You have to make an extra call to gather tags (describe-tags). So, I did a quick workaround to group nodes by "prefix". If you follow a certain naming convention on your AWS resources, that will help solving your issue too. But, I'm still having issues with some resources that are generated automatically and does not follow naming conventions. Tags will be really nice.

**Comment by 0xdabbad00** [comment]
There is a `--collapse-by-tag` filter currently in the `prepare` step. This was added for cases where people are handling auto-scaling on their own, without using AWS's proper autoscaling groups, so they have dozens or more EC2s that are all exactly the same. That could help you here.

It sounds like you might have say an account with different applications or environments in it that have some tag, let's say it's tagged `prod` and `staging`, so you want to see just the `prod` resources.  I worry that if we limit your view to only the `prod` resources, it will hide communication paths that may exist to the `staging` resources that you would want to be aware of.  Maybe you don't care?  Can you explain your use case a bit more and desired output so I can make sure I get this right?

I'm also thinking about filtering down to a single VPC or Subnet, which could help as well.

**Comment by rafarost** [comment]
In my case I'm actually using the --collapse-by-tag feature. It helped me a lot on ec2 instances. But, I do have a similar scenario on RDS instances, which makes really hard to read the graph. Since I know that all the instances have the same network setup, and all of them follow a naming convention, I did an implementation to group them by a prefix. That worked nice for me. But, group by tag would be much more elegant. Gather tags will also be nice for ELB's. For instance, when you have all ELB's being auto-generated, with random names. Their tags will actually have more information regarding the services behind them. So, If you have an option to name them according an specific tag, would be really nice to make the graph more readable.

**Comment by 0xdabbad00** [comment]
I ticketed https://github.com/duo-labs/cloudmapper/issues/5 to provide some customization about what is displayed, but we'll focus on the filtering in this ticket.

It sounds like you need support for the `--collapse-by-tag` to be added to more resources @rafarost, such as RDS instances.  I'll look into what I can do there.

Also, for any others reading this.  Another option, which is not great, but could help for unique use cases, would be to edit your .json files yourself.  For example, for the demo files you could run:
```
cat describe-instances.json | jq '.Reservations[].Instances[] | select(.Tags[].Value=="Web2")'
```
You would need to add back in the parents, but that would show only instances with a tag value of `Web2`  (Somehow you could probably maintain the parents with `jq` but I'm not sure how).

**Comment by ngerstle** [comment]
The ability to toggle a view (or collapse-by-tag, which I just saw) for some nodes would be fantastic.
Many of the accounts/services I've tried mapping have underlying non-amazon infrastructure services (consul/prometheus/logging) that clutter and obscure all the other network connections because they connect to all nodes. 
Being able to toggle (especially in browser) visibility for all resources with a given tag would be good, and dynamically selecting displayed tags would be fantastic.

**Comment by GauravThukral** [comment]
I am trying to use --collapse-by-tag when using python cloudmapper.py prepare. In AWS tag there is key Environment which has enviornment name . How can i use same in command.

python cloudmapper.py prepare --account-name tmonpe --collapse-by-tag Environment:Environment name

**Comment by 0xdabbad00** [comment]
Filtering by tags resolved in #343
